File: libs/src/model-scenarios/d2/modules/scenarios/api/HistoryStatisticsApiEndpoint.java
package d2.modules.scenarios.api;


import d2.modules.scenarios.model.Scenario;
import d2.modules.scenarios.model.Subject;
import ic.id.Mapper;
import ic.network.SocketAddress;
import ic.network.http.*;
import ic.storage.BindingStorage;
import ic.struct.collection.Collection;
import ic.struct.collection.ConvertCollection;
import ic.struct.collection.JoinCollection;
import ic.throwables.NotExists;
import ic.throwables.UnableToParse;
import org.json.JSONObject;

import static d2.modules.scenarios.model.History.getHistory;
import static d2.modules.scenarios.model.Scenario.getScenariosMapper;
import static d2.modules.scenarios.model.Subject.getSubjectsMapper;
import static ic.date.DateFormat.formatDateOrThrow;
import static ic.json.JsonArrays.toJsonArray;
import static ic.throwables.Skip.SKIP;


@Deprecated
public abstract class HistoryStatisticsApiEndpoint extends ProtectedHttpEndpoint {

	@Override protected String getName() { return "history_statistics"; }

	protected abstract Collection<String> getUserIds();
	protected abstract BindingStorage getStorage();

	@Override protected HttpResponse implementSafeEndpoint(SocketAddress socketAddress, HttpRequest request) throws UnableToParse {

		final JSONObject responseJson = new JSONObject();

		responseJson.put("status", "SUCCESS");

		final BindingStorage storage = getStorage();

		responseJson.put("history", toJsonArray(

			new JoinCollection<>(
				new ConvertCollection<>(
					getUserIds(),
					userId -> {
						final Mapper<Subject, String> subjectsMapper = getSubjectsMapper(storage);
						return new JoinCollection<>(
							new ConvertCollection<>(
								subjectsMapper.getIds(),
								subjectId -> {
									final Subject subject = subjectsMapper.getItem(subjectId);
									final Mapper<Scenario, String> scenariosMapper = getScenariosMapper(storage, subjectId);
									return new JoinCollection<>(
										new ConvertCollection<>(
											scenariosMapper.getIds(),
											scenarioId -> {
												final Scenario scenario = scenariosMapper.getItem(scenarioId);
												return new ConvertCollection<>(
													getHistory(storage, userId, subjectId, scenarioId).getItems(),
													item -> {
														final JSONObject itemJson = new JSONObject();
														itemJson.put("userId", userId);
														itemJson.put("subject", subject.getTitle());
														itemJson.put("scenario", scenario.getTitle());
														try {
															itemJson.put("date", formatDateOrThrow(item.date, "yyyy.MM.dd HH:mm:ss"));
														} catch (NotExists notExists) { throw SKIP; }
														try {
															itemJson.put("blockText", scenario.getBlockByIdOrThrow(item.blockId).content);
														} catch (NotExists notExists) {
															itemJson.put("blockText", "blockId: " + item.blockId);
														}
														itemJson.put("way", item.way);
														itemJson.put("score", item.score);
														return itemJson;
													}
												);
											}
										)
									);
								}
							)
						);
					}
				)
			),

			json -> json

		));

		return new JsonResponse() {
			@Override protected Object getJson() { return responseJson; }
		};

	}

}

File: libs/src/model-scenarios/d2/modules/scenarios/api/statistics/HistoryStatisticsApiEndpoint.kt
@file:Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")


package d2.modules.scenarios.api.statistics


import d2.modules.scenarios.model.History
import ic.network.SocketAddress
import ic.network.http.HttpRequest
import ic.network.http.HttpResponse
import ic.network.http.JsonResponse
import ic.network.http.ProtectedHttpEndpoint
import ic.storage.BindingStorage
import ic.struct.collection.Collection
import ic.struct.collection.ConvertCollection
import ic.struct.collection.JoinCollection
import ic.throwables.NotExists
import ic.throwables.UnableToParse
import org.json.JSONObject

import d2.modules.scenarios.model.History.getHistory
import d2.modules.scenarios.model.Scenario.getScenariosMapper
import d2.modules.scenarios.model.Subject.getSubjectsMapper
import ic.date.DateFormat.formatDateOrThrow
import ic.interfaces.getter.Getter1
import ic.json.JsonArrays.toJsonArray
import ic.throwables.Skip.SKIP


abstract class HistoryStatisticsApiEndpoint : ProtectedHttpEndpoint() { override val name get() = "history"

	protected abstract val userIds: Collection<String>
	protected abstract val storage: BindingStorage

	@Throws(UnableToParse::class)
	override fun implementSafeEndpoint(socketAddress: SocketAddress, request: HttpRequest): HttpResponse {

		val responseJson = JSONObject()

		responseJson.put("status", "SUCCESS")

		val storage = storage

		responseJson.put("history", toJsonArray(

			JoinCollection<JSONObject>(
				ConvertCollection<Collection<JSONObject>, String>(userIds) { userId ->
					val subjectsMapper = getSubjectsMapper(storage)
					JoinCollection<JSONObject>(
						ConvertCollection<Collection<JSONObject>, String>(subjectsMapper.ids) { subjectId ->
							val subject = subjectsMapper[subjectId]
							val scenariosMapper = getScenariosMapper(storage, subjectId)
							JoinCollection<JSONObject>(
								ConvertCollection<Collection<JSONObject>, String>(scenariosMapper.ids) { scenarioId ->
									val scenario = scenariosMapper[scenarioId]
									ConvertCollection(
										getHistory(storage, userId, subjectId, scenarioId).items,
										object : Getter1<JSONObject, History.Item> {
											var previousItem : History.Item? = null
											override fun get (item : History.Item) : JSONObject {
												System.out.println("$userId $subjectId $scenarioId")
												if (previousItem == null) {
													previousItem = item
													throw SKIP
												}
												val itemJson = JSONObject()
												itemJson.put("userId", userId)
												itemJson.put("subject", subject.title)
												itemJson.put("scenario", scenario.title)
												try {
													val block = scenario.getBlockByIdOrThrow(previousItem!!.blockId)
													itemJson.put("blockText", block.content)
													itemJson.put(
														"way",
														(
															if (item.way == -1) "Back" else
															if (block.keyboardType.startsWith("AB")) {
																if (item.way == 0) "A" else
																if (item.way == 1) "B" else
																if (item.way == 2) "C" else
																if (item.way == 3) "D"
																else "?"
															}
															else "Next"
														)
													)
												} catch (notExists: NotExists) {
													itemJson.put("blockText", "blockId: " + previousItem!!.blockId)
													itemJson.put("way", item.way.toString())
												}
												try {
													itemJson.put("date", formatDateOrThrow(item.date, "yyyy.MM.dd HH:mm:ss"))
												} catch (notExists: NotExists) { throw SKIP }
												itemJson.put("score", item.score)
												previousItem = item
												return itemJson
											}
										}
									)
								}
							)
						}
					)
				}
			)

		) { json -> json })

		return JsonResponse.Final(responseJson)

	}

}
